File: Languages/python/algorithms-dsa/leetcode/trees/maximum_depth.py
"""
LeetCode #104 - Maximum Depth of Binary Tree
Topic: Trees
Difficulty: Easy

Find max depth using recursive DFS.

Time Complexity: O(n)
Space Complexity: O(h) where h = tree height
"""


class TreeNode:
    def __init__(self, val: int = 0, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


def max_depth(root: TreeNode | None) -> int:
    if not root:
        return 0
    return 1 + max(max_depth(root.left), max_depth(root.right))


if __name__ == "__main__":
    tree = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    assert max_depth(tree) == 3
    assert max_depth(TreeNode(1, None, TreeNode(2))) == 2
    assert max_depth(None) == 0
    print("All tests passed!")

File: Languages/go/algorithms-dsa/leetcode/trees/maximum_depth.go
/*
LeetCode #104: Maximum Depth of Binary Tree
Topic: Trees
Difficulty: Easy

Given the root of a binary tree, return its maximum depth. Maximum depth
is the number of nodes along the longest path from root to leaf.

Approach: Recursive DFS. Depth = 1 + max(left depth, right depth).

Time: O(n), Space: O(h) where h is height
*/
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left := maxDepth(root.Left)
	right := maxDepth(root.Right)
	if left > right {
		return left + 1
	}
	return right + 1
}

func main() {
	tree := &TreeNode{3,
		&TreeNode{9, nil, nil},
		&TreeNode{20, &TreeNode{15, nil, nil}, &TreeNode{7, nil, nil}},
	}
	if maxDepth(tree) != 3 {
		fmt.Println("FAIL Test 1")
	} else {
		fmt.Println("PASS Test 1")
	}

	if maxDepth(nil) != 0 {
		fmt.Println("FAIL Test 2")
	} else {
		fmt.Println("PASS Test 2")
	}

	if maxDepth(&TreeNode{Val: 1}) != 1 {
		fmt.Println("FAIL Test 3")
	} else {
		fmt.Println("PASS Test 3")
	}
}
